File: packages/flutter_fuels/lib/model/input.dart
import 'package:flutter_fuels/utils/hex_utils.dart';
import 'package:flutter_fuels/utils/json_utils.dart';

abstract class Input {
  const Input();

  static Input fromJson(Map<String, dynamic> jsonInput) {
    try {
      int inputType = jsonInput['type'];
      switch (inputType) {
        case 0:
          return InputCoin.fromJson(jsonInput);
        case 1:
          return InputContract.fromJson(jsonInput);
        case 2:
          return InputMessage.fromJson(jsonInput);
        default:
          return UnknownInput(raw: jsonInput);
      }
    } catch (e) {
      return UnparsedInput(raw: jsonInput, err: e);
    }
  }
}

class UnknownInput extends Input {
  final Map<String, dynamic> raw;

  const UnknownInput({required this.raw});
}

class UnparsedInput extends Input {
  final Map<String, dynamic> raw;
  final dynamic err;

  const UnparsedInput({required this.raw, required this.err});
}

class InputCoin extends Input {
  final String owner;
  final BigInt amount;
  final String assetId;
  final int witnessIndex;

  const InputCoin(
      {required this.owner,
      required this.amount,
      required this.assetId,
      required this.witnessIndex});

  factory InputCoin.fromJson(Map<String, dynamic> data) {
    return InputCoin(
      owner: addHexPrefix(data['owner']),
      amount: parseBigInt(data['amount']),
      assetId: addHexPrefix(data['assetId']),
      witnessIndex: data['witnessIndex'],
    );
  }
}

class InputContract extends Input {
  final String contractID;

  const InputContract({required this.contractID});

  factory InputContract.fromJson(Map<String, dynamic> data) {
    return InputContract(contractID: addHexPrefix(data['contractID']));
  }
}

class InputMessage extends Input {
  final BigInt amount;
  final String sender;
  final String recipient;
  final int witnessIndex;

  const InputMessage({
    required this.amount,
    required this.sender,
    required this.recipient,
    required this.witnessIndex,
  });

  factory InputMessage.fromJson(Map<String, dynamic> data) {
    return InputMessage(
        amount: parseBigInt(data['amount']),
        sender: addHexPrefix(data['sender']),
        recipient: addHexPrefix(data['recipient']),
        witnessIndex: data['witnessIndex']);
  }
}

File: packages/fuels/native/src/model/input.rs
pub enum Input {
    InputCoin {
        owner: String,
        amount: u64,
        asset_id: String,
        witness_index: Option<u16>,
    },
    InputContract {
        contract_id: String,
    },
    InputMessage {
        amount: u64,
        sender: String,
        recipient: String,
        witness_index: Option<u16>,
    },
    UnknownInput,
}

impl From<&fuel_tx::Input> for Input {
    fn from(model: &fuel_tx::Input) -> Self {
        match model {
            fuel_tx::Input::CoinSigned(coin) => { Input::InputCoin { owner: coin.owner.to_string(), amount: coin.amount, asset_id: coin.asset_id.to_string(), witness_index: Some(coin.witness_index) } }
            fuel_tx::Input::CoinPredicate(coin) => { Input::InputCoin { owner: coin.owner.to_string(), amount: coin.amount, asset_id: coin.asset_id.to_string(), witness_index: None } }
            fuel_tx::Input::Contract(contract) => { Input::InputContract { contract_id: contract.contract_id.to_string() } }
            fuel_tx::Input::MessageCoinSigned(msg) => { Input::InputMessage { amount: msg.amount, sender: msg.sender.to_string(), recipient: msg.recipient.to_string(), witness_index: Some(msg.witness_index) } }
            fuel_tx::Input::MessageCoinPredicate(msg) => { Input::InputMessage { amount: msg.amount, sender: msg.sender.to_string(), recipient: msg.recipient.to_string(), witness_index: None } }
            fuel_tx::Input::MessageDataSigned(msg) => { Input::InputMessage { amount: msg.amount, sender: msg.sender.to_string(), recipient: msg.recipient.to_string(), witness_index: Some(msg.witness_index) } }
            fuel_tx::Input::MessageDataPredicate(msg) => { Input::InputMessage { amount: msg.amount, sender: msg.sender.to_string(), recipient: msg.recipient.to_string(), witness_index: None } }
        }
    }
}
